enumerable_methods.rb

```diff
@@ -0,0 +1,155 @@
+# rubocop:disable Metrics/ModuleLength
+module Enumerable
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/PerceivedComplexity
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    i = 0
+    while i < size
+      if is_a? Hash
+        yield(keys[i], self[keys[i]])
+        return self if i == size - 1
+      elsif is_a? Array
+        yield self[i]
+        return self if i == size - 1
+      elsif is_a? Range
+        yield to_a[i]
+        return self if i == size - 1
+      end
+      i += 1
+    end
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    i = 0
+    while i < size
+      if is_a? Hash
+        yield(keys[i], self[keys[i]])
+        return self if i == size - 1
+      elsif is_a? Array
+        yield(self[i], i)
+        return self if i == size - 1
+      elsif is_a? Range
+        yield to_a[i], i
+        return self if i == size - 1
+      end
+      i += 1
+    end
+  end
```

Comment:
`my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- Your my_each_with_index is not interacting with with the hash in the following form. 
```ruby
hash = { a: 1, b: 2, c: 3, d: 4, e: 5 }
my_each_with_index_output = ''
hash.my_each_with_index { |num, idx| my_each_with_index_output += "Num: #{num}, idx: #{idx}\n" }
p my_each_with_index_output
my_each_with_index_output = ''
hash.each_with_index { |num, idx| my_each_with_index_output += "Num: #{num}, idx: #{idx}\n" }
p my_each_with_index_output
```
You can play with the solution given above to make it workable.